subsys/bluetooth/mesh/light_ctrl_reg_spec.c

```diff
@@ -45,21 +38,64 @@ static void reg_step(struct k_work *work)
 		ki = spec_reg->reg.cfg.ki.down;
 	}
 
-	spec_reg->i += (input * ki) * ((float)REG_INT / (float)MSEC_PER_SEC);
-	spec_reg->i = CLAMP(spec_reg->i, 0, UINT16_MAX);
+	return (struct reg_terms){
+		.i = ((input) * (ki) * ((float)REG_INT / (float)MSEC_PER_SEC)),
+		.p = input * kp,
+	};
+}
+
+static void reg_step(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_mesh_light_ctrl_reg_spec *spec_reg = CONTAINER_OF(
+		dwork, struct bt_mesh_light_ctrl_reg_spec, timer);
+	struct reg_terms reg_terms;
+
+	if (!spec_reg->enabled) {
+		/* The regulator might be disabled asynchronously. */
+		return;
+	}
+
+	k_work_reschedule(&spec_reg->timer, K_MSEC(REG_INT));
+
+	reg_terms = reg_terms_calc(spec_reg);
+	spec_reg->i += reg_terms.i;
+
+	if (spec_reg->i >= 0) {
+		/* Drop the negative flag as soon as the internal sum becomes positive. */
+		spec_reg->neg = false;
+	}
+
+	if (!spec_reg->neg) {
+		spec_reg->i = CLAMP(spec_reg->i, 0, UINT16_MAX);
+	}
 
-	float p = input * kp;
-	float output = spec_reg->i + p;
+	float output = spec_reg->i + reg_terms.p;
 
 	spec_reg->reg.updated(&spec_reg->reg, output);
 }
 
-void bt_mesh_light_ctrl_reg_spec_start(struct bt_mesh_light_ctrl_reg *reg)
+static void internal_sum_recover(struct bt_mesh_light_ctrl_reg_spec *spec_reg, uint16_t lightness)
+{
+	struct reg_terms reg_terms;
+
+	reg_terms = reg_terms_calc(spec_reg);
+
+	/* Recalculate the intenal sum so that it is equal to the passed lightness level at the
```

Comment:
```suggestion
	/* Recalculate the internal sum so that it is equal to the passed lightness level at the
```
I'm actually amazed I failed to catch this in the previous round LOL